src/aurelia-event-aggregator.ts

```diff
@@ -0,0 +1,168 @@
+import * as LogManager from 'aurelia-logging';
+
+const logger = LogManager.getLogger('event-aggregator');
+
+export type Constructor<T = object> = new (...args: any[]) => T;
+export type EventCallback = (data?: any, event?: string) => any;
+
+export class Handler {
+  constructor(public messageType: Constructor, public callback: EventCallback) {}
+
+  public handle(message: object) {
+    if (message instanceof this.messageType) {
+      this.callback.call(null, message);
+    }
+  }
+}
+
+function invokeCallback(callback: EventCallback, data: any, event: string) {
+  try {
+    callback(data, event);
+  } catch (e) {
+    logger.error(e);
+  }
+}
+
+function invokeHandler(handler: Handler, data: object) {
+  try {
+    handler.handle(data);
+  } catch (e) {
+    logger.error(e);
+  }
+}
+
+/**
+ * Represents a disposable subsciption to an EventAggregator event.
+ */
+export interface Subscription {
+  /**
+   * Disposes the subscription.
+   */
+  dispose(): void;
+}
+
+/**
+ * Enables loosely coupled publish/subscribe messaging.
+ */
+export class EventAggregator {
+  /**
+   * Creates an instance of the EventAggregator class.
+   */
+  constructor(
+    private eventLookup: {[event: string]: EventCallback[]} = {},
+    private messageHandlers: Handler[] = []
+  ) {}
+
+  /**
+   * Publishes a message.
+   * @param event The event or channel to publish to.
+   * @param data The data to publish on the channel.
+   */
+  public publish(event: string | object, data?: any): void {
+    let subscribers;
+    let i;
+
+    if (!event) {
+      throw new Error('Event was invalid.');
+    }
+
+    if (typeof event === 'string') {
+      subscribers = this.eventLookup[event];
+      if (subscribers) {
+        subscribers = subscribers.slice();
+        i = subscribers.length;
+
+        while (i--) {
+          invokeCallback(subscribers[i], data, event);
+        }
+      }
+    } else {
+      subscribers = this.messageHandlers.slice();
+      i = subscribers.length;
+
+      while (i--) {
+        invokeHandler(subscribers[i], event);
+      }
+    }
+  }
+
+  /**
+   * Subscribes to a message channel or message type.
+   * @param event The event channel or event data type.
+   * @param callback The callback to be invoked when when the specified message is published.
+   */
+  public subscribe(event: string | Constructor, callback: EventCallback): Subscription {
```

Comment:
What about having an overload for the subscribe methods that are generic, so that you can get some typing around the callbacks?

Reply:
Good idea. I'll get it done.